### PR TITLE
TUSD Clipper Callee

### PR DESCRIPTION
**Pull request by talbaneth**

### CODE REVIEW

File: src/TUSDCurveCallee.sol [hunk 1]
```diff
@@ -0,0 +1,122 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+// Copyright (C) 2021 Dai Foundation
```
Comment (julienmartinlevrai) [inline src/TUSDCurveCallee.sol:2]: 2022

File: src/TUSDCurveCallee.sol [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+pragma solidity >=0.6.12;
```
Comment (julienmartinlevrai) [inline src/TUSDCurveCallee.sol:17]: Remove the >=. It is best practice to fix the compiler version

File: src/test/TUSDCurveCallee.t.sol [hunk 3]
```diff
@@ -0,0 +1,332 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+// Copyright (C) 2021 Dai Foundation
```
Comment (julienmartinlevrai) [inline src/test/TUSDCurveCallee.t.sol:2]: 2022

File: src/test/TUSDCurveCallee.t.sol [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+pragma solidity ^0.6.12;
```
Comment (julienmartinlevrai) [inline src/test/TUSDCurveCallee.t.sol:17]: Remove the ^

File: src/test/TUSDCurveCallee.t.sol [hunk 5]
```diff
@@ -0,0 +1,332 @@
...
+        Hevm(hevm).warp(block.timestamp + tail / 2);
+        Clipper(clipper).take({
+            id:   id,
+            amt:  amt,
+            max:  type(uint256).max,
+            who:  address(callee),
+            data: data
+        });
+
+        address dai = Chainlog(chainlog).getAddress("MCD_DAI");
```
Comment (julienmartinlevrai) [inline src/test/TUSDCurveCallee.t.sol:215]: This can be moved to `setUp`

File: src/test/TUSDCurveCallee.t.sol [hunk 6]
```diff
@@ -0,0 +1,332 @@
...
+        uint256 amt = 20000 * WAD;
+        newAuction(amt);
+        bytes memory data = abi.encode(
+            address(this),
+            address(gemJoin),
+            uint256(0)
+        );
+        Hevm(hevm).warp(block.timestamp + tail / 2);
+        address osm = Chainlog(chainlog).getAddress("PIP_TUSD");
+        Hevm(hevm).store({
```
Comment (julienmartinlevrai) [inline src/test/TUSDCurveCallee.t.sol:293]: This can go in`setUp()`

File: src/test/TUSDCurveCallee.t.sol [hunk 7]
```diff
@@ -0,0 +1,332 @@
...
+
+    function testFail_maxPrice() public {
+        uint256 amt = 20000 * WAD;
+        newAuction(amt);
+        bytes memory data = abi.encode(
+            address(this),
+            address(gemJoin),
+            uint256(0)
+        );
+        Hevm(hevm).warp(block.timestamp + tail / 5);
```
Comment (julienmartinlevrai) [inline src/test/TUSDCurveCallee.t.sol:316]: What is the reasoning behind the choice of this specific number ? Maybe add a comment to explain.

**Review by julienmartinlevrai — COMMENTED** [review]
This is a first pass. I will make a more thorough review soon.

The test structure is very clean.

**Comment by talbaneth** [comment]
@julienmartinlevrai @godsflaw, please wait with further reviewing the tests, doing changes to align with the new proposed parameters.

File: src/TUSDCurveCallee.sol (hunk 1 above)
Comment (talbaneth) [reply src/TUSDCurveCallee.sol:2]: Done

File: src/TUSDCurveCallee.sol (hunk 2 above)
Comment (talbaneth) [reply src/TUSDCurveCallee.sol:17]: Done

File: src/test/TUSDCurveCallee.t.sol (hunk 3 above)
Comment (talbaneth) [reply src/test/TUSDCurveCallee.t.sol:2]: Done

File: src/test/TUSDCurveCallee.t.sol (hunk 4 above)
Comment (talbaneth) [reply src/test/TUSDCurveCallee.t.sol:17]: Done

File: src/test/TUSDCurveCallee.t.sol (hunk 5 above)
Comment (talbaneth) [reply src/test/TUSDCurveCallee.t.sol:215]: Done

File: src/test/TUSDCurveCallee.t.sol (hunk 7 above)
Comment (talbaneth) [reply src/test/TUSDCurveCallee.t.sol:316]: No real reasoning as this was ported from the test of WstETH.
Added the following comment:
"Make sure price goes down enough to make up for the Curve slippage'

File: src/test/TUSDCurveCallee.t.sol (hunk 6 above)
Comment (talbaneth) [reply src/test/TUSDCurveCallee.t.sol:293]: Moved the osm address retreiving to setup as all the chainlog queries.
Left the taking of ownership here as it is only used in this test.

**Comment by talbaneth** [comment]
@julienmartinlevrai @godsflaw tests are ready to review again

**Review by julienmartinlevrai — COMMENTED** [review]
In general it looks good, I left some minor comments. I haven’t looked at the tests yet

File: src/test/TUSDCurveCallee.t.sol [hunk 8]
```diff
@@ -0,0 +1,383 @@
...
+        takeOwnership(spotter);
+        Spot(spotter).file("TUSD-A", bytes32("mat"), 150 * RAY / 100);
+        renounceOwnership(spotter);
+        Spot(spotter).poke("TUSD-A");
+
+        id = Dog(dog).bark("TUSD-A", address(this), address(this));
+    }
+
+    function test_baseline() public {
+        uint256 amt = 20000 * WAD;
```
Comment (julienmartinlevrai) [inline src/test/TUSDCurveCallee.t.sol:255]:
```suggestion
        uint256 amt = 20_000 * WAD;
```

File: src/test/TUSDCurveCallee.t.sol [hunk 9]
```diff
@@ -0,0 +1,383 @@
...
+        Clipper(clipper).file("chip", 0);
+
+        // Set tip to 500 (kick all at start to achieve linear price discount, a bit higher value so that we guarantee kicking them all - ideally PE helps here)
+        Clipper(clipper).file("tip", 500 * RAD);
+
+        renounceOwnership(clipper);
+        renounceOwnership(dog);
+    }
+
+    function newAuction(uint256 amt) internal {
```
Comment (julienmartinlevrai) [inline src/test/TUSDCurveCallee.t.sol:223]:
```suggestion
    function newAuction(uint256 amt) internal returns (uint256 id) {
```

File: src/test/TUSDCurveCallee.t.sol [hunk 10]
```diff
@@ -0,0 +1,383 @@
...
+        Clipper(clipper).take({
+            id:   id,
+            amt:  amt,
+            max:  type(uint256).max,
+            who:  address(callee),
+            data: data
+        });
+
+        assertEq(Token(dai).balanceOf(address(this)), 0);
+        assertEq(Token(tusd).balanceOf(address(this)), 0);
```
Comment (julienmartinlevrai) [inline src/test/TUSDCurveCallee.t.sol:272]: I am not sure I understand these checks. I would suggest to check `clipper.count() == prevCount - 1` instead

File: src/test/TUSDCurveCallee.t.sol [hunk 11]
```diff
@@ -0,0 +1,383 @@
...
+        uint256 amt = 26_000_000 * WAD;
+        newAuction(amt);
+        bytes memory data = abi.encode(
+            address(123),
+            address(gemJoin),
+            uint256(minProfit)
+        );
+
+        // Jump ahead to when price goes down enough to make up for the slippage
+        Hevm(hevm).warp(block.timestamp + tail * 9000 / 10000);
```
Comment (julienmartinlevrai) [inline src/test/TUSDCurveCallee.t.sol:326]:
```suggestion
        Hevm(hevm).warp(block.timestamp + tail * 9 / 10);
```

File: src/test/TUSDCurveCallee.t.sol [hunk 12]
```diff
@@ -0,0 +1,383 @@
...
+            amt:  amt,
+            max:  type(uint256).max,
+            who:  address(callee),
+            data: data
+        });
+        assertGe(Token(dai).balanceOf(address(123)), minProfit);
+    }
+
+    function test_maxPrice() public {
+        uint256 amt = 20000 * WAD;
```
Comment (julienmartinlevrai) [inline src/test/TUSDCurveCallee.t.sol:338]:
```suggestion
        uint256 amt = 20_000 * WAD;
```

File: src/test/TUSDCurveCallee.t.sol [hunk 13]
```diff
@@ -0,0 +1,383 @@
...
+            id:   id,
+            amt:  amt,
+            max:  max,
+            who:  address(callee),
+            data: data
+        });
+    }
+
+    function test_tailPrice() public {
+        uint256 amt = 20000 * WAD;
```
Comment (julienmartinlevrai) [inline src/test/TUSDCurveCallee.t.sol:364]:
```suggestion
        uint256 amt = 20_000 * WAD;
```

**Review by julienmartinlevrai — COMMENTED** [review]
Left minor comments for the tests, in general it looks good

File: src/test/TUSDCurveCallee.t.sol (hunk 10 above)
Comment (talbaneth) [reply src/test/TUSDCurveCallee.t.sol:281]: Added count checls.

**Review by julienmartinlevrai — APPROVED** [review]
All my comments have been addressed and all tests pass on commit `21eebed`

LGTM

File: scripts/deploy-mainnet.sh [hunk 14]
```diff
@@ -23,8 +23,13 @@ CurveLpTokenUniv3Callee=$(dapp create CurveLpTokenUniv3Callee \
   0x9759A6Ac90977b93B58547b4A71c78317f391A28 \
   0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2)
 
+TUSDCurveCallee=$(dapp create TUSDCurveCallee \
+  0xEcd5e75AFb02eFa118AF914515D6521aaBd189F1 \
```
Comment (julienmartinlevrai) [inline scripts/deploy-mainnet.sh:27]: Although this seems to be the correct TUSD Curve Metapool, I was unable to verify its authenticity as it was deployed by a [Metapool factory](0x0959158b6040D32d04c301A72CBFD6b39E21c9AE) that doesn’t match the [Curve documentation](https://curve.readthedocs.io/ref-addresses.html#metapool-factory)

Comment (talbaneth) [reply scripts/deploy-mainnet.sh:27]: That factory seems to be an old one, found it in one of their contract:
https://github.com/curvefi/curve-factory/blob/master/contracts/Factory.vy#L112

You can see Curve presents the pool address here - https://curve.fi/contracts.
Look for "tusd".

Comment (julienmartinlevrai) [reply scripts/deploy-mainnet.sh:27]: correct.

**Review by julienmartinlevrai — APPROVED** [review]

**Comment by talbaneth** [comment]
TUSD was offboarder successfully, no need to merge this, closing.

**Pull request closed without merging**